orangecontrib/text/bagofowords.py

```diff
@@ -0,0 +1,78 @@
+import warnings
+
+import numpy as np
+
+from gensim import corpora, matutils
+from gensim.models.tfidfmodel import TfidfModel
+
+from orangecontrib.text.preprocess import Preprocessor
+
+
+class BagOfWords():
+
+    def __init__(self, progress_callback=None, error_callback=None):
+        self.progress_callback = progress_callback
+        self.error_callback = error_callback
+        self.vocabulary = None
+
+    def __call__(self, corpus, use_tfidf=False):
+        if corpus is None:
+            raise ValueError(
+                    'Cannot compute Bag of Words without an input corpus.'
+            )
+
+        try:
+            has_tokens = not (corpus.tokens is None)
+        except:
+            has_tokens = False
+
+        if not has_tokens:  # Perform default pre-processing.
+            preprocessor = Preprocessor()
+            corpus = preprocessor(corpus)
+
+        self.check_progress()  # Step 1
+
+        try:
+            dictionary = corpora.Dictionary(
+                    corpus.tokens,
+                    prune_at=float('Inf')
+            )
+            self.vocabulary = dictionary
+
+            # Term frequencies.
+            bag_of_words = [dictionary.doc2bow(i) for i in corpus.tokens]
+
+            self.check_progress()  # Step 2
+
+            if use_tfidf:
+                tfidf_model = TfidfModel(bag_of_words)
+                bag_of_words = [
```

Comment:
No need to iterate, could be done like `bag_of_words = tfidf_model[bag_of_words]`.